### PR TITLE
Address extra `FutureOr` from _isolates_io.dart.

### PR DESCRIPTION
**Pull request by lrhn**
## Description

The `_spawn` function takes a `_IsolateConfiguration<Q, FutureOr<R>>` configuration, then tries to assign `await configuration.apply()` to `R`. However the `apply` function returns `FutureOr<FutureOr<R>>` which becomes `FutureOr<R>` when awaited.
The analyzer has so far not caught this problem, but I'm fixing that, and this code fails in the fixed analyzer.

I added an extra `await` on the result from `apply`. You may want to consider whether the code can be rewritten to avoid this.

The type checking with only the one `await` used to work in the compilers because of implicit down-casts (`FutureOr<R>` is assignable to `R` in plain Dart through an implicit downcast). The analyzer failed to recognize this as an implicit downcast, so didn't trigger the no-implicit-downcasts lint that you otherwise use.

It actually *is* a downcast which just happens to have always succeeded.

## Tests

The following test failure following a Dart analyzer change discovered this bug:
https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket.appspot.com/8891023110533266960/+/steps/analyze_flutter/0/stdout

### CODE REVIEW

**Comment by goderbauer** [comment]
Looks like the analyzer is unhappy with this change.

**Comment by lrhn** [comment]
That fix was too naïve. OK, I'll just add the extra `await` that your existing type requires. Then you can consider whether there is a better rewrite.

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/_isolates_io.dart]: nit: There's an extra whitespace at the end of this line that the analyzer is complaining about.

**Review by goderbauer — COMMENTED** [review]
I seem to remember that `Future<Future<R>>` would resolve to R when awaited. This is not true for nested `FutureOr`s?

**Comment by lrhn** [comment]
> I seem to remember that `Future<Future<R>>` would resolve to R when awaited. This is not true for nested `FutureOr`s?

It's not true for anything. That behavior was not type-safe, so it went away with Dart 2.0.
The analyzer has not completely changed the analysis around `FutureOr` to match yet, which is the bug I'm trying to fix in https://dart-review.googlesource.com/c/sdk/+/121846, which was then blocked on this code which depended on the incorrect analysis (and implicit downcasts at run-time).

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/_isolates_io.dart]: make this final?

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by lrhn** [comment]
Fixed now, ready to be merged when tree is open.

**Review by Piinks — COMMENTED** [review]
Hi @lrhn! Welcome 🎉  and thank you for the contribution! This will merge automatically once the build is green. :)

**Pull request merged**
